### PR TITLE
Remove geometry check for st_contain and st_within

### PR DESCRIPTION
**Pull request by WangCHX**
Instructions:
1. The PR has to be tagged with at least one of the following labels (*):
   1. `enhancement`

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/util/GeoColumnUtils.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Optional;
+import org.apache.pinot.common.request.context.ExpressionContext;
+
+
+public final class GeoColumnUtils {
+
+  private GeoColumnUtils() {
+  }
+
+  public static Optional<String> getColumnFromArgument(ExpressionContext argument) {
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/util/GeoColumnUtils.java:31]: add some comments

also, i think this handles first level func only

Comment (WangCHX) [reply pinot-core/src/main/java/org/apache/pinot/core/util/GeoColumnUtils.java:37]: got it. added comments and also handled multiple level.

Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/util/GeoColumnUtils.java:37]: Thanks, it's possible for other nested functions such as some geometry manipulation.

**Review by yupeng9 — APPROVED** [review]

**Comment by yupeng9** [comment]
plz take a look at the tests

**Review by Jackie-Jiang — CHANGES_REQUESTED** [review]
This is not the correct fix. The filter operator relies on the H3 index on the column. If the column is within an expression, the result will be wrong. E.g. `ST_DISTANCE(geoCol, 'point')` is not the same as `ST_DISTANCE(myFunc(geoCol), 'point')`, and we cannot use the same index because the index is not applied to the `myFunc(geoCol)`

**Comment by WangCHX** [comment]
> This is not the correct fix. The filter operator relies on the H3 index on the column. If the column is within an expression, the result will be wrong. E.g. `ST_DISTANCE(geoCol, 'point')` is not the same as `ST_DISTANCE(myFunc(geoCol), 'point')`, and we cannot use the same index because the index is not applied to the `myFunc(geoCol)`

is it ok for me to add constraints that the function only can be `toSphericalGeography` and `toGeometry` right now?

**Comment by Jackie-Jiang** [comment]
> > This is not the correct fix. The filter operator relies on the H3 index on the column. If the column is within an expression, the result will be wrong. E.g. `ST_DISTANCE(geoCol, 'point')` is not the same as `ST_DISTANCE(myFunc(geoCol), 'point')`, and we cannot use the same index because the index is not applied to the `myFunc(geoCol)`
> 
> is it ok for me to add constraints that the function only can be `toSphericalGeography` and `toGeometry` right now?

Why do you need to do `toSphericalGeography` or `toGeometry` on a column with H3 index? IIUC, H3 index implies the underlying points are geography, and `ST_DISTANCE(toGeometry(geoCol), 'point')` should not use H3 index.

**Comment by WangCHX** [comment]
> > > This is not the correct fix. The filter operator relies on the H3 index on the column. If the column is within an expression, the result will be wrong. E.g. `ST_DISTANCE(geoCol, 'point')` is not the same as `ST_DISTANCE(myFunc(geoCol), 'point')`, and we cannot use the same index because the index is not applied to the `myFunc(geoCol)`
> > 
> > 
> > is it ok for me to add constraints that the function only can be `toSphericalGeography` and `toGeometry` right now?
> 
> Why do you need to do `toSphericalGeography` or `toGeometry` on a column with H3 index? IIUC, H3 index implies the underlying points are geography, and `ST_DISTANCE(toGeometry(geoCol), 'point')` should not use H3 index.

I think the underlying is a point. H3 index only take coordinate. so it can be applied to  both geography and geometry. Also [ST_Contains](https://docs.pinot.apache.org/configuration-reference/functions/stcontains) only can be applied to  `Geometry`. If i want to use ST_DISTANCE and ST_Contains at the same time, seems I need to create two columns.

**Comment by Jackie-Jiang** [comment]
@yupeng9 Can we apply `ST_Contains` to the geography objects? I think it should work on both geometry and geography

**Comment by yupeng9** [comment]
current implementation of `ST_Contains`  is for Geometry only
https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java

**Comment by Jackie-Jiang** [comment]
> current implementation of `ST_Contains` is for Geometry only https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java

@yupeng9 Understood, but why does it not apply to the geography objects?

**Comment by yupeng9** [comment]
> > current implementation of `ST_Contains` is for Geometry only https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java
> 
> @yupeng9 Understood, but why does it not apply to the geography objects?

Different implementation, and I didn't scope it in the inital impl

**Comment by Jackie-Jiang** [comment]
> > > current implementation of `ST_Contains` is for Geometry only https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java
> > 
> > 
> > @yupeng9 Understood, but why does it not apply to the geography objects?
> 
> Different implementation, and I didn't scope it in the inital impl

@yupeng9 Is H3 index always based on the geography objects? If so, we cannot use H3 index for ST_Contains because that will give wrong result (mixing geometry calculation with geography calculation).
Another option is to allow using geometry contains on geography objects because they should give close estimation on small areas. It won't work for large areas though.

**Comment by yupeng9** [comment]
> > > > current implementation of `ST_Contains` is for Geometry only https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java
> > > 
> > > 
> > > @yupeng9 Understood, but why does it not apply to the geography objects?
> > 
> > 
> > Different implementation, and I didn't scope it in the inital impl
> 
> @yupeng9 Is H3 index always based on the geography objects? If so, we cannot use H3 index for ST_Contains because that will give wrong result (mixing geometry calculation with geography calculation). Another option is to allow using geometry contains on geography objects because they should give close estimation on small areas. It won't work for large areas though.

Yes, H3 is always based on geography. And yes, we can do approximation for small area.

**Comment by Jackie-Jiang** [comment]
Then my suggestion is to remove the geometry check in the `StContainsFunction`, and we should document the behavior of using geometry contains for geography objects, which can give good estimation on small areas. We should also add a TODO to support accurate contains on geography objects in the future.

**Comment by WangCHX** [comment]
>

**Comment by WangCHX** [comment]
> Then my suggestion is to remove the geometry check in the `StContainsFunction`, and we should document the behavior of using geometry contains for geography objects, which can give good estimation on small areas. We should also add a TODO to support accurate contains on geography objects in the future.

thanks Jackie and Yupeng. just curious why it is not accurate? My use case is checking all the coordinates inside a city polygon (like SF). is it small area or big area for a city like SF?

**Comment by Jackie-Jiang** [comment]
> > Then my suggestion is to remove the geometry check in the `StContainsFunction`, and we should document the behavior of using geometry contains for geography objects, which can give good estimation on small areas. We should also add a TODO to support accurate contains on geography objects in the future.
> 
> thanks Jackie and Yupeng. just curious why it is not accurate? My use case is checking all the coordinates inside a city polygon (like SF). is it small area or big area for a city like SF?

IMO a single city on the earth sphere can be considered small area.
I can see one exception area around longitude of -180/180, where geography and geometry will give very different result (in geography, -179.9 to 179.9 is a short distance, but in geometry that is a whole circle of the sphere). In other areas treating geography as geometry should give close approximation.
@yupeng9 What do you think if we allow `ST_Contains` and `ST_Within` on geography but use the geometry algorithm?

**Comment by yupeng9** [comment]
@Jackie-Jiang is spot on. There are corner cases around the boundary, and thats the risk you have face with the approximation. Alternatively, you can explore the ST_Contains implementation for geography and add the proper impl to Pinot

**Comment by Jackie-Jiang** [comment]
I think we can merge this and update our documentation to warn about the behavior so that H3 index can be used for now. The geography support can be added in a separate PR. @yupeng9 wdyt?

**Comment by yupeng9** [comment]
I'm fine with it. @WangCHX Please highlight this and put a warning in the user doc

**Review by Jackie-Jiang — APPROVED** [review]

**Comment by WangCHX** [comment]
> I'm fine with it. @WangCHX Please highlight this and put a warning in the user doc

sure. will add that.

**Comment by WangCHX** [comment]
doc: https://github.com/pinot-contrib/pinot-docs/pull/92

**Pull request merged**